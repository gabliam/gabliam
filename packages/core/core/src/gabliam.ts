/* eslint-disable no-await-in-loop */
import { ExpressionParser } from '@gabliam/expression';
import d from 'debug';
import _ from 'lodash';
import { callInstance, toPromise } from './common';
import { APP_CONFIG, CORE_CONFIG, TYPE, VALUE_EXTRACTOR } from './constants';
import { Container, createContainer } from './container';
import {
  ConfigRegistry,
  GabliamAddPlugin,
  GabliamConfig,
  PreDestroyRegistry,
  ValueRegistry,
} from './interfaces';
import { LoaderConfig, LoaderModule } from './loaders';
import {
  Bean,
  BeforeCreate,
  Init,
  OnMissingBean,
  PreDestroy,
} from './metadatas';
import { PluginList } from './plugin-list';
import { reflection } from './reflection';
import { Registry } from './registry';
import { configureValueExtractor } from './value-extractor';
import { getGabliamConfig } from './gabliam-options';

const debug = d('Gabliam:core');

/**
 * Gabliam
 */
export class Gabliam {
  public container: Container = createContainer();

  public config: object;

  /**
   * Registry
   */
  public registry: Registry = new Registry();

  /**
   * Module loader
   */
  public loaderModule: LoaderModule = new LoaderModule();

  /**
   * Config loader
   */
  public loaderConfig: LoaderConfig = new LoaderConfig();

  /**
   * Plugin list
   */
  public pluginList: PluginList = new PluginList();

  public options: GabliamConfig;

  /**
   * Constructor
   * @param  {GabliamConfig|string} options?
   */
  constructor(options?: Partial<GabliamConfig> | string) {
    this.options = getGabliamConfig(options);
    /**
     * @TODO move in building phase
     */
    this.container.bind(CORE_CONFIG).toConstantValue(this.options);
  }

  /**
   * Add a plugin
   * @param  {GabliamPluginConstructor} plugin
   * @returns Gabliam
   */
  public addPlugin(plugin: GabliamAddPlugin): Gabliam {
    this.pluginList.add(plugin);
    return this;
  }

  /**
   * Add any plugins
   * @param  {GabliamPluginConstructor[]} ...plugins
   * @returns Gabliam
   */
  public addPlugins(...plugins: Array<GabliamAddPlugin>): Gabliam {
    for (const plugin of plugins) {
      this.addPlugin(plugin);
    }
    return this;
  }

  /**
   * Build gabliam
   * @returns Promise
   */
  public async build(): Promise<Gabliam> {
    this.pluginList.sort();

    /**
     * Load config file
     */
    await this._initializeConfig();

    /**
     * Loading phase
     */
    this.registry.addRegistry(
      this.loaderModule.load(this.options.scanPath, this.pluginList.plugins),
    );

    /**
     * Binding phase
     */
    await this._bind();

    /**
     * Config phase
     */
    await this._loadConfig();

    for (const plugin of this.pluginList.pluginsWithBuild) {
      await toPromise(plugin.build(this.container, this.registry));
    }

    return this;
  }

  /**
   * Build and start gabliam application
   * @returns Promise
   */
  async buildAndStart(): Promise<Gabliam> {
    await this.build();
    await this.start();
    return this;
  }

  /**
   * Stop and destroy gabliam application
   * @returns Promise
   */
  async stopAndDestroy(): Promise<Gabliam> {
    await this.stop();
    await this.destroy();
    return this;
  }

  /**
   * Start gabliam application
   *
   * call all plugin.start
   * @returns Promise
   */
  async start(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithStart) {
      await toPromise(plugin.start(this.container, this.registry));
    }

    return this;
  }

  /**
   * Stop gabliam application
   * call all plugin.stop
   *
   * @returns Promise
   */
  async stop(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithStop) {
      await toPromise(plugin.stop(this.container, this.registry));
    }

    return this;
  }

  /**
   * Destroy gabliam application
   * call all plugin.destroy
   *
   * @returns Promise
   */
  async destroy(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithDestroy) {
      await toPromise(plugin.destroy(this.container, this.registry));
    }

    const values = this.registry.get<PreDestroyRegistry>(TYPE.PreDestroy);

    const instanceToDestroy = async (
      value: ValueRegistry<PreDestroyRegistry>,
    ) => {
      if (value.options) {
        const instance = this.container.get(value.id);
        for (const preDestroy of value.options.preDestroys) {
          await callInstance(instance, preDestroy);
        }
      }
    };

    const p = [];
    for (const value of values) {
      p.push(instanceToDestroy(value));
    }

    await Promise.all(p);

    return this;
  }

  /**
   * Load config file and bind result in APP_CONFIG
   */
  private async _initializeConfig() {
    this.config = await this.loaderConfig.load(this.options.config);
    const config = this.config;
    this.container.bind(APP_CONFIG).toConstantValue(config);
    this.container
      .bind(VALUE_EXTRACTOR)
      .toConstantValue(configureValueExtractor(this.container));
    this.container
      .bind(ExpressionParser)
      .toConstantValue(new ExpressionParser(config));
  }

  /**
   * Binding phase
   * Binding of config and service classes.
   * call all plugin.bind
   */
  private async _bind() {
    debug('_bind');

    this.registry
      .getAllAutoBind()
      .forEach(({ id, target }) =>
        this.container.bind(id).to(target).inSingletonScope(),
      );

    for (const plugin of this.pluginList.pluginsWithBind) {
      await toPromise(plugin.bind(this.container, this.registry));
    }
  }

  /**
   * Config phase
   */
  private async _loadConfig() {
    debug('_loadConfig');

    // Get all config classes in registry
    let configsRegistry = this.registry.get<ConfigRegistry>(TYPE.Config);

    debug('configsRegistry', configsRegistry);
    if (configsRegistry) {
      configsRegistry = _.sortBy(configsRegistry, 'options.order');

      for (const { id: configId } of configsRegistry) {
        // Get config instance
        const confInstance = this.container.get<object>(configId);
        const ctor = confInstance.constructor;

        // get all bean metadata in config classes
        const beanMetadatas = reflection.propMetadataOfDecorator<Bean>(
          ctor,
          Bean,
        );

        // get on missing bean metadata
        const onMissingBeanMetadatas = reflection.propMetadataOfDecorator<OnMissingBean>(
          ctor,
          OnMissingBean,
        );

        const beforeCreateMetas = Object.keys(
          reflection.propMetadataOfDecorator(ctor, BeforeCreate),
        );

        const initMetadas = Object.keys(
          reflection.propMetadataOfDecorator(ctor, Init),
        );

        // call all beforeCreate method if exist
        if (Array.isArray(beforeCreateMetas)) {
          // No promise.all and await because order of beans are important
          for (const metada of beforeCreateMetas) {
            await callInstance(confInstance, metada);
          }
        }

        // If config has bean metadata
        if (Object.keys(beanMetadatas).length) {
          // No promise.all and await because order of beans are important
          for (const [key, beans] of Object.entries(beanMetadatas)) {
            const onMissingBeans = onMissingBeanMetadatas[key] || [];

            // by default all bean are missing
            let allMissing = true;

            // if there are onMissingBeans config, check if bean exist or no
            for (const onMissingBean of onMissingBeans) {
              try {
                this.container.getAll(onMissingBean.id);
                allMissing = false;
                // eslint-disable-next-line no-empty
              } catch {}
            }

            // if all beans are missing, so we create all beans
            if (allMissing) {
              for (const { id } of beans) {
                const bean = await callInstance(confInstance, key);
                this.container.bind(id).toConstantValue(bean);
                if (bean === undefined || bean.constructor === undefined) {
                  // eslint-disable-next-line no-continue
                  continue;
                }

                const preDestroys = Object.keys(
                  reflection.propMetadataOfDecorator(
                    bean.constructor,
                    PreDestroy,
                  ),
                );

                // bean can return undefined or can be a constant value
                if (preDestroys.length) {
                  this.registry.add(TYPE.PreDestroy, <ValueRegistry>{
                    id,
                    target: bean,
                    options: {
                      preDestroys,
                    },
                  });
                }
              }
            }
          }
        }

        if (Array.isArray(initMetadas)) {
          // No promise.all and await because order of beans are important
          for (const metada of initMetadas) {
            await callInstance(confInstance, metada);
          }
        }

        for (const plugin of this.pluginList.pluginWithConfig) {
          await toPromise(
            plugin.config(this.container, this.registry, confInstance),
          );
        }
      }
    }
    debug('_loadConfig end');
  }
}
