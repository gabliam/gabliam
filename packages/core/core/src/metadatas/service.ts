/* eslint-disable @typescript-eslint/no-redeclare */
import { injectable, interfaces } from 'inversify';
import { METADATA_KEY, TYPE } from '../constants';
import { makeDecorator } from '../decorator';
import { Register } from './register';

/**
 * Type of the `Service` decorator / constructor function.
 */
export interface ServiceDecorator {
  /**
   * Decorator that marks a class as an Gabliam Service and provides configuration
   * metadata that determines how the config should be processed,
   * instantiated.
   *
   * @usageNotes
   *
   * ```typescript
   * @Service()
   * class GretterService {
   *    constructor(private gretter: Gretter) {
   *       gretter.greet();
   *    }
   * }
   *
   *  class Gretter {
   *      constructor(private name:string){};
   *      greet() {
   *          return `Hello ${this.name} !`;
   *      }
   *  }
   *
   * @Config()
   * class SampleConfig {
   *      @Bean(Gretter)
   *      createGretter() {
   *          return new Gretter('David');
   *      }
   * }
   * ```
   */
  (name?: string): ClassDecorator;

  /**
   * see the `@Service` decorator.
   */
  new (name?: string): any;
}

/**
 * `Service` decorator and metadata.
 */
export interface Service {
  /**
   * Name of Service
   */
  name?: string;
}

export const Service: ServiceDecorator = makeDecorator(
  METADATA_KEY.service,
  (name?: string): Service => ({ name }),
  (cls, annotationInstance: Service) => {
    const id: interfaces.ServiceIdentifier<any> = annotationInstance.name
      ? annotationInstance.name
      : cls;
    injectable()(cls);
    Register({ type: TYPE.Service, id })(cls);
  },
);
