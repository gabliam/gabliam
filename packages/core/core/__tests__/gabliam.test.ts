// tslint:disable:one-line
// tslint:disable:no-unused-expression
import {
  Gabliam,
  APP_CONFIG,
  CORE_CONFIG,
  Container,
  Registry,
  Config,
  Plugin,
  GabliamPlugin,
  Init,
  Bean,
  OnMissingBean,
  BeforeCreate,
} from '../src';
import * as path from 'path';
import { TestService } from './fixtures/gabliam/service';
import { DbConfig } from './fixtures/gabliam/db-config';
import * as sinon from 'sinon';
import { GabliamTest } from '../src/testing/gabliam';

test('gabliam instance', async () => {
  const gab = new Gabliam({
    scanPath: path.resolve(__dirname, './fixtures/gabliam'),
    config: path.resolve(__dirname, './fixtures/gabliam/config'),
  });
  await gab.buildAndStart();
  // @todo write a guid serializer
  // expect(gab).toMatchSnapshot();
  expect(gab.container.get(APP_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(CORE_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(TestService)).toMatchSnapshot();
  expect(gab.container.get(DbConfig)).toMatchSnapshot();
  await gab.destroy();
});

test('gabliam instance with default config', async () => {
  const gab = new Gabliam();
  expect(gab.options).toEqual({ config: undefined });
  await gab.destroy();
});

test('gabliam instance with config path', async () => {
  const gab = new Gabliam(path.resolve(__dirname, './fixtures/gabliam/config'));
  expect(gab.options).toMatchSnapshot();
  await gab.destroy();
});

test('gabliam instance with path', async () => {
  const gab = new Gabliam({
    scanPath: path.resolve(__dirname, './fixtures/gabliam'),
    config: path.resolve(__dirname, './fixtures/gabliam/config'),
  });
  await gab.buildAndStart();
  // @todo write a guid serializer
  // expect(gab).toMatchSnapshot();
  expect(gab.container.get(APP_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(CORE_CONFIG)).toMatchSnapshot();
  expect(gab.container.get(TestService)).toMatchSnapshot();
  expect(gab.container.get(DbConfig)).toMatchSnapshot();
  await gab.destroy();
});

describe('test plugin', () => {
  @Plugin()
  class PluginTest implements GabliamPlugin {
    build(container: Container, registry: Registry): void {}

    bind(container: Container, registry: Registry): void {}

    config(container: Container, registry: Registry, confInstance: any): void {}

    start(container: Container, registry: Registry): Promise<void> {
      return Promise.resolve();
    }

    stop(container: Container, registry: Registry): Promise<void> {
      return Promise.resolve();
    }

    destroy(container: Container, registry: Registry): Promise<void> {
      return Promise.resolve();
    }
  }

  const build = sinon.spy(PluginTest.prototype, 'build');
  const bind = sinon.spy(PluginTest.prototype, 'bind');
  const config = sinon.spy(PluginTest.prototype, 'config');
  const start = sinon.spy(PluginTest.prototype, 'start');
  const stop = sinon.spy(PluginTest.prototype, 'stop');
  const destroy = sinon.spy(PluginTest.prototype, 'destroy');
  let gab: Gabliam;

  describe('with config', () => {
    let res = '';
    beforeAll(() => {
      const g = new GabliamTest(new Gabliam().addPlugin(PluginTest));
      gab = g.gab;

      @Config(300)
      class Conf {
        constructor() {
          res += 'Conf';
        }
      }
      @Config(200)
      class Conf2 {
        constructor() {
          res += 'Conf2';
        }
      }

      g.addClass(Conf);
      g.addClass(Conf2);
    });

    afterAll(() => {
      build.resetHistory();
      bind.resetHistory();
      config.resetHistory();
      start.resetHistory();
      stop.resetHistory();
      destroy.resetHistory();
    });

    test('gabliam build', async () => {
      await gab.build();
      expect(build.calledOnce).toBe(true);
      expect(bind.calledOnce).toBe(true);
      expect(config.callCount).toMatchSnapshot();
      expect(res).toMatchSnapshot();
    });

    test('gabliam start', async () => {
      await gab.start();
      expect(start.calledOnce).toBe(true);
    });

    test('gabliam stop', async () => {
      await gab.stop();
      expect(stop.calledOnce).toBe(true);
    });

    test('gabliam destroy', async () => {
      await gab.destroy();
      expect(destroy.calledOnce).toBe(true);
    });
  });

  describe('without config', () => {
    beforeAll(() => {
      gab = new Gabliam({
        scanPath: path.resolve(__dirname, './fixtures/gabliam'),
        config: path.resolve(__dirname, './fixtures/gabliam/config'),
      });
      gab.addPlugin(PluginTest);
    });

    afterAll(() => {
      build.resetHistory();
      bind.resetHistory();
      config.resetHistory();
      start.resetHistory();
      stop.resetHistory();
      destroy.resetHistory();
    });

    test('gabliam build', async () => {
      await gab.build();
      expect(build.calledOnce).toBe(true);
      expect(bind.calledOnce).toBe(true);
      expect(config.calledOnce).toBe(true);
    });

    test('gabliam start', async () => {
      await gab.start();
      expect(start.calledOnce).toBe(true);
    });

    test('gabliam stop', async () => {
      await gab.stop();
      expect(stop.calledOnce).toBe(true);
    });

    test('gabliam destroy', async () => {
      await gab.destroy();
      expect(destroy.calledOnce).toBe(true);
    });
  });
});

test('should fail with bad Plugin', () => {
  class BadPlugin {}
  const g = new GabliamTest();
  expect(() => {
    g.gab.addPlugin(BadPlugin);
  }).toThrowError();
});

test('ok with Plugin', async () => {
  @Plugin()
  class Plugin1 {}

  @Plugin({ dependencies: ['Plugin1'] })
  class Plugin2 {}

  @Plugin({ dependencies: [{ name: 'Plugin2', order: 'before' }] })
  class Plugin3 {}
  const g = new GabliamTest();
  g.gab.addPlugins(Plugin1, Plugin2, Plugin3);
  await expect(g.build()).resolves;
});

test('ok with Plugin 2', async () => {
  @Plugin()
  class Plugin1 {}

  @Plugin({ dependencies: ['Plugin1'] })
  class Plugin2 {}

  @Plugin({ dependencies: [{ name: 'Plugin2', order: 'after' }] })
  class Plugin3 {}

  @Plugin({
    dependencies: [{ name: 'Plugin3', order: 'after' }],
  })
  class Plugin4 {}
  const g = new GabliamTest();
  g.gab.addPlugin(Plugin4);
  g.gab.addPlugin(Plugin3);
  g.gab.addPlugin(Plugin1);
  g.gab.addPlugin(Plugin2);
  await expect(g.build()).resolves;
});

test('ok with Plugin 3', async () => {
  @Plugin()
  class Plugin1 {}

  @Plugin({ dependencies: [Plugin1] })
  class Plugin2 {}

  @Plugin({ dependencies: [{ name: Plugin2, order: 'before' }] })
  class Plugin3 {}
  const g = new GabliamTest();
  g.gab.addPlugins(Plugin3);
  await expect(g.build()).resolves;
});

test('should fail with Plugin with bad deps', async () => {
  @Plugin({ dependencies: ['BadDeps'] })
  class BadPlugin {}
  const g = new GabliamTest();
  g.gab.addPlugin(BadPlugin);
  await expect(g.build()).rejects.toMatchSnapshot();
});

test('plugin findByName', () => {
  @Plugin()
  class BadPlugin {}
  const g = new GabliamTest();
  g.gab.addPlugin(BadPlugin);
  expect(g.gab.pluginList.findByName('BadPlugin')).toMatchSnapshot();
});

test('test init', async () => {
  let res = '';
  @Config(300)
  class Conf {
    constructor() {
      res += '|Conf';
    }

    @Init()
    init() {
      res += '|initConf';
    }
  }
  @Config(200)
  class Conf2 {
    constructor() {
      res += '|Conf2';
    }

    @Init()
    init() {
      res += '|initConf2';
    }
  }
  const g = new GabliamTest();
  g.addClass(Conf);
  g.addClass(Conf2);
  await g.build();
  expect(res).toMatchSnapshot();
});

test('@onMissingBean', async () => {
  let res = '';
  @Config(300)
  class Conf {
    constructor() {
      res += '|constructConf';
    }

    @OnMissingBean('conf1')
    @Bean('conf1')
    conf1() {
      res += '|conf1';
    }
  }
  @Config(200)
  class Conf2 {
    constructor() {
      res += '|constructConf2';
    }

    @OnMissingBean('conf1')
    @Bean('conf2')
    conf2() {
      res += '|Conf2';
    }
  }
  const g = new GabliamTest();
  g.addClass(Conf);
  g.addClass(Conf2);
  await g.build();
  expect(res).toMatchSnapshot();
});

test('@onMissingBean 2', async () => {
  let res = '';
  @Config(200)
  class Conf {
    constructor() {
      res += '|constructConf';
    }

    @OnMissingBean('conf1')
    @Bean('conf1')
    conf1() {
      res += '|conf1';
    }
  }
  @Config(300)
  class Conf2 {
    constructor() {
      res += '|constructConf2';
    }

    @OnMissingBean('conf1')
    @Bean('conf2')
    conf2() {
      res += '|Conf2';
    }
  }
  const g = new GabliamTest();
  g.addClass(Conf);
  g.addClass(Conf2);
  await g.build();
  expect(res).toMatchSnapshot();
});

test('@onMissingBean with multiple missing but just one is missing', async () => {
  let res = '';
  @Config(200)
  class Conf {
    constructor() {
      res += '|constructConf';
    }

    @OnMissingBean('conf1')
    @Bean('conf1')
    conf1() {
      res += '|conf1';
    }
  }
  @Config(300)
  class Conf2 {
    constructor() {
      res += '|constructConf2';
    }

    @OnMissingBean('conf3')
    @OnMissingBean('conf1')
    @Bean('conf2')
    conf2() {
      res += '|Conf2';
    }
  }
  const g = new GabliamTest();
  g.addClass(Conf);
  g.addClass(Conf2);
  await g.build();
  expect(res).toMatchSnapshot();
});

test('test BeforeCreate', async () => {
  let res = '';
  @Config(300)
  class Conf {
    constructor() {
      res += '|Conf';
    }

    @BeforeCreate()
    beforeCreate() {
      res += '|BeforeCreateConf';
    }
  }
  @Config(200)
  class Conf2 {
    constructor() {
      res += '|Conf2';
    }

    @BeforeCreate()
    beforeCreate() {
      res += '|BeforeCreateConf2';
    }
  }
  const g = new GabliamTest();
  g.addClass(Conf);
  g.addClass(Conf2);
  await g.build();
  expect(res).toMatchSnapshot();
});

test('test BeforeCreate & init', async () => {
  let res = '';
  @Config(300)
  class Conf {
    constructor() {
      res += '|Conf';
    }
    @Init()
    init() {
      res += '|InitCreateConf';
    }

    @Bean('create')
    create() {
      res += '|create';
      return 'create';
    }

    @BeforeCreate()
    beforeCreate() {
      res += '|BeforeCreateConf';
    }
  }
  @Config(200)
  class Conf2 {
    constructor() {
      res += '|Conf2';
    }

    @BeforeCreate()
    beforeCreate() {
      res += '|BeforeCreateConf2';
    }

    @Bean('create2')
    create() {
      res += '|create2';
      return 'create2';
    }

    @Init()
    init() {
      res += '|InitCreateConf2';
    }
  }
  const g = new GabliamTest();
  g.addClass(Conf);
  g.addClass(Conf2);
  await g.build();
  g.gab.container.get('create');
  g.gab.container.get('create2');
  expect(res).toMatchSnapshot();
});
