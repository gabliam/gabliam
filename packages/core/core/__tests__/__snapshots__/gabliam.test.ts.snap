// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@onMissingBean 1`] = `"|constructConf2|Conf2|constructConf|conf1"`;

exports[`@onMissingBean 2 1`] = `"|constructConf|conf1|constructConf2"`;

exports[`@onMissingBean with multiple missing but just one is missing 1`] = `"|constructConf|conf1|constructConf2"`;

exports[`gabliam instance 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`gabliam instance 2`] = `
Object {
  "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/gabliam/config",
  "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/gabliam",
}
`;

exports[`gabliam instance 3`] = `
TestService {
  "dbConfig": DbConfig {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`gabliam instance 4`] = `
DbConfig {
  "db": "test",
  "host": "127.0.0.1",
}
`;

exports[`gabliam instance with config path 1`] = `
Object {
  "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/gabliam/config",
}
`;

exports[`gabliam instance with path 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`gabliam instance with path 2`] = `
Object {
  "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/gabliam/config",
  "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/gabliam",
}
`;

exports[`gabliam instance with path 3`] = `
TestService {
  "dbConfig": DbConfig {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`gabliam instance with path 4`] = `
DbConfig {
  "db": "test",
  "host": "127.0.0.1",
}
`;

exports[`plugin findByName 1`] = `
Object {
  "beforeAll": false,
  "dependencies": Array [],
  "name": "BadPlugin",
  "plugin": BadPlugin {},
}
`;

exports[`should fail with Plugin with bad deps 1`] = `[PluginDependencyIsMissingError: The plugin BadPlugin need the plugin BadDeps. Try to install it: npm install BadDeps --save or yarn add BadDeps]`;

exports[`test BeforeCreate & init 1`] = `"|Conf2|BeforeCreateConf2|create2|InitCreateConf2|Conf|BeforeCreateConf|create|InitCreateConf"`;

exports[`test BeforeCreate 1`] = `"|Conf2|BeforeCreateConf2|Conf|BeforeCreateConf"`;

exports[`test init 1`] = `"|Conf2|initConf2|Conf|initConf"`;

exports[`test plugin with config gabliam build 1`] = `2`;

exports[`test plugin with config gabliam build 2`] = `"Conf2Conf"`;
